### ISSUE TITLE
Flow-sensitive typing / typestate

### ISSUE THREAD

**Issue by Ericson2314** [python/mypy#6477]
To adequately handle mutation, the type of the variable cannot be everywhere constant but instead allowed to vary with respect to position in control flow. The trick is tackle control flow statically, with something like regions/lifetime of the control flow, so we can still assign types to (variable, region)s, statically.

In this literature this seems to go by both Flow-sensitive typing (more recent) and typestate (older, typically cfg-sensative refinement types of cfg-agnostic base types).

It seems mypy can handle this in a few cases through ad-hoc means, but there's no general story on this front. A small but interesting example I don't think that is handled yet is:
```python
x: Optional[T] = ...
if x is None:
  raise ...
x.method_on_t(...)
```

I assume the thing to do is 1. extend the internal notion of type assignments to be CF-aware (the typed langauge itself), and then 2. work on inference for this richer language. Once 1 is done, but before 2 is, a `typing.assert` analogous to `typing.cast` would be the ideal primitive to do this manually.

**Comment by JukkaL** [comment]
I believe that your example works correctly -- mypy doesn't generate an error. Mypy already has pretty powerful flow sensitive type inference (but obviously it's not perfect). If you'd like to improve things, a good first step would be to come up with more (complete and tested :-) examples where mypy currently doesn't do the right thing. After that we can discuss how important it would be to support those examples and what an implementation might look like.

**Comment by Ericson2314** [comment]
OK that's all very good to here! That would just leave the typing assertion part, I think? Can one do a `foo: Ty` "statement" with an already-existing variable?

**Comment by JelleZijlstra** [comment]
`assert isinstance(foo, Ty)` would do that if I understand you correctly.

**Comment by ilevkivskyi** [comment]
... as well as `if isinstance(foo, Ty)` (depending on what exactly do you mean)

**Comment by Ericson2314** [comment]
Great! Happy to be wrong here.

**Comment by Ericson2314** [comment]
Is there a "static assert" which has no runtime check? Or if I `typing.cast` and don't use the result does mypy use that as an hint about the argument's type?

**Comment by JelleZijlstra** [comment]
I haven't tried it, but I'd assume `if TYPE_CHECKING: assert isinstance(...)` works.

**Comment by Ericson2314** [comment]
Great!

**Comment by Ericson2314** [comment]
I think this is enough good news to close it for real now. Thanks you all.
